README.md

<!--- README.md is generated from README.Rmd. Please edit that file -->
Optimal Project Prioritization Protocol
=======================================

[![lifecycle](https://img.shields.io/badge/Lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental) [![Travis Build Status](https://img.shields.io/travis/prioritizr/optimalppp/master.svg?label=Linux%20%26%20Mac%20OSX)](https://travis-ci.org/prioritizr/optimalppp) [![AppVeyor Build Status](https://img.shields.io/appveyor/ci/jeffreyhanson/optimalppp/master.svg?label=Windows)](https://ci.appveyor.com/project/jeffreyhanson/optimalppp) [![Coverage Status](https://codecov.io/github/prioritizr/optimalppp/coverage.svg?branch=master)](https://codecov.io/github/prioritizr/optimalppp?branch=master) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/optimalppp)](https://CRAN.R-project.org/package=optimalppp)

**This package is still under development and not ready for use. Please do not use this package yet.**

The *optimalppp R* package provides methods for prioritizing funding of conservation projects using the 'Protect Prioritization Protocol'. A range of methods are provided for identifying priority projects for funding. These include exact algorithm solvers which can identify optimal solutions, and also stingy heuristic algorithms that have conventionally been used to identify solutions. This package also provides the functionality to visualize how well solutions maintain biodiversity.

Installation
------------

The latest development version can be installed using the following code. Please note that you will need install the *ggtree* package from Bioconductor since it is not available on [The Comprehensive R Archive Network](https://cran.r-project.org/).

``` r
if (!require(devtools))
  install.packages("devtools")
devtools::install_github("prioritizr/optimalppp")
if (!require(ggtree))
  devtools::install_bioc("ggtree")
```

Usage
-----

Here we will provide a short example showing how the *optimalppp R* package can be used to prioritize funding for conservation projects. To start off, we will set the seed for the random number generator to ensure you get the same results as shown here, set some default behavior for the *R* session, load the *optimalppp R* package, and load the *ggtree R* package to plot phylogenetic trees.

``` r
set.seed(500)
library(optimalppp)
library(ggtree)
```

Now we will load some data sets that are distributed with the package. First, we will load the `sim_tree` object. This object describes the evolutionary relationships between 5 simulated species (named S1, S2, S3, ..., S5). The length of each phylogenetic branch corresponds to millions of years of evolutionary history, and the last common ancestor for all these species occurred approximately 2 million years ago.

``` r
# load data
data(sim_tree)

# plot tree
ggtree(sim_tree) +
geom_tiplab(size = 2.5) +
geom_treescale(width = 2, x = 0, offset = 0.1)
```

<img src="man/figures/README-unnamed-chunk-6-1.png" style="display: block; margin: auto;" />

Next, we will load the `sim_project_data` object. This object stores information about various conservation projects in a tabular format (i.e. `tibble`). Each row corresponds to a different project, and each column describes various properties associated with the projects. Importantly, the `"name"` column contains the name of each project, the `"cost"` column contains the cost of each project, the `"success"` column denotes the probability of each project succeeding if it is funded, and the `"S1"`--`"SN"` columns show the enhanced probability of each species persisting if the project is funded. It also contains additional columns for customizing the solutions, but we will ignore them for now. Note that the last project---the `"baseline_project"`---has a zero cost and represents the baseline probability of each species persisting if no other project is funded. Finally, though most projects in this example directly relate to a single species, you can input projects that directly affect the persistence of multiple species.

``` r
# load data
data(sim_project_data)

# print table
head(as.data.frame(sim_project_data))
```

    ##               name      cost   success locked_in locked_out        S1
    ## 1       S1_project  94.39929 0.8470486     FALSE      FALSE 0.8022048
    ## 2       S2_project 100.99137 0.9694998     FALSE      FALSE 0.0000000
    ## 3       S3_project 103.22583 0.7323494     FALSE       TRUE 0.0000000
    ## 4       S4_project  99.24274 0.9792034     FALSE      FALSE 0.0000000
    ## 5       S5_project  99.90791 0.7142838      TRUE      FALSE 0.0000000
    ## 6 baseline_project   0.00000 1.0000000     FALSE      FALSE 0.2933155
    ##          S2         S3        S4        S5
    ## 1 0.0000000 0.00000000 0.0000000 0.0000000
    ## 2 0.8079388 0.00000000 0.0000000 0.0000000
    ## 3 0.0000000 0.67841372 0.0000000 0.0000000
    ## 4 0.0000000 0.00000000 0.7034556 0.0000000
    ## 5 0.0000000 0.00000000 0.0000000 0.6569473
    ## 6 0.3883986 0.01203652 0.1952440 0.1237303

Let us assume that our resources are limited such that we can only spend, at most, $200 on funding conservation projects. In other words, our budget is capped at $200. Now, given the project data (`sim_project_data`), the species' evolutionary relationships (`sim_tree`), and this budget (`200`), So, let's cut to the chase and find an optimal solution.

``` r
# solve problem
s1 <- ppp_exact_solution(x = sim_project_data, tree = sim_tree,
                         budget = 200, project_column_name = "name",
                         cost_column_name = "cost",
                         success_column_name = "success")

# print solution
head(as.data.frame(s1))
```

    ##   solution objective budget    cost optimal method S1_project S2_project
    ## 1        1  3.045741    200 193.642    TRUE  exact       TRUE      FALSE
    ##   S3_project S4_project S5_project baseline_project
    ## 1      FALSE       TRUE      FALSE             TRUE

The object `s1` contains the solution and also various statistics associated with the solution in a tabular format (i.e. `tibble`). Here, each row corresponds to a different solution. Specifically, the `"solution"` column contains an identifier for the solution (this is useful for methods that output multiple solutions), the `"objective"` column contains the objective value (i.e. the expected phylogenetic diversity, Faith 2008), the `"budget"` column stores the budget used for generating the solution, the `"cost"` column stores the cost of the solution, the `"optimal"` column indicates if the solution is known to be optimal (`NA` values mean the optimality is unknown), and the `"method"` column contains the name of the method used to generate the solution. The remaining columns (`"S1_project"`, `"S2_project"`, `"S3_project"`, ..., `"S50_project"`, and `"baseline_project"`) indicate if a given project was prioritized for funding in the solution or not.

Here, the objective value (in the `"objective"` column) denotes the amount of evolutionary history that is expected to persist (i.e. 3.046 million years). Put simply, solutions that are expected to result in better conservation outcomes will be associated with a greater objective value. Since tabular data can be difficult to intuit, let's visualize how well this solution would maintain the different branches in the phylogeny. Note that species which receive any funding are denoted with an asterisk.

``` r
# visualize solution
ppp_plot(sim_project_data, sim_tree, s1, project_column_name = "name",
         cost_column_name = "cost", success_column_name = "success") +
geom_treescale(width = 2, x = 0, offset = 0.1)
```

<img src="man/figures/README-unnamed-chunk-9-1.png" style="display: block; margin: auto;" />

This has just been a taster of the *optimalppp R* package. For more information, see the [package vignette](https://prioritizr.github.io/optimalppp/articles/optimalppp.html).

Citation
--------

Please use the following citation to cite the *optimalppp R* package in publications:

**This package is still under development and not ready for use. Please do not use this package yet.**

Hanson JO, Schuster R, Strimas-Mackey M, Bennett J, (2018). optimalppp: Optimal Project Prioritization Protocol. R package version 0.0.0.3. Available at <https://github.com/prioritizr/optimalppp>.


NEWS.md
# optimalppp 0.0.0.4

- Progress on solving the "right" problem.

# optimalppp 0.0.0.3

- This package currently solves the wrong problem, reverting the package
  state back to experimental.

# optimalppp 0.0.0.2

- Initial GitHub release where everything appears to work.

# optimalppp 0.0.0.1

- Initial commit.


R/ppp_phylo_plot.R
#' @include internal.R
NULL

#' Plot a 'Project Prioritization Protocol' solution
#'
#' Create a plot displaying a phylogenetic tree (phylogram) to visualize a
#' solution to the 'Project Prioritization Protocol' problem
#' (Joseph, Maloney & Possingham 2009). In this plot, each
#' phylogenetic branch is colored according to probability
#' that it is expected to persist into the future (based on Faith 2008).
#' Additionally, species that benefit from at least a single funded project
#' with a non-zero cost are denoted with an asterisk symbol.
#'
#' @inheritParams help
#'
#' @param n \code{integer} solution to display. This should argument
#'   correspond to a row number in the argument to \code{solution}. Defaults
#'   to \code{1} such that the solution in the first row of the argument
#'   to \code{solution} is plotted.
#'
#' @param asterisk_hjust \code{numeric} horizontal adjustment parameter to
#'   manually align the asterisks in the plot. Defaults to \code{0.007}.
#'   Increasing this parameter will shift the asterisks further left towards
#'   the species labels.
#'
#' @details This function requires the \pkg{ggtree} (Yu \emph{et al.} 2017).
#'   Since this package is distributed exclusively
#'   through \href{https://bioconductor.org}{Bioconductor}, and is not
#'   available on the
#'   \href{https://cran.r-project.org/}{Comprehensive R Archive Network},
#'   please execute the following command to install it:
#'   \code{source("https://bioconductor.org/biocLite.R");biocLite("ggtree")}.
#'   If the installation process fails, please consult the package's \href{https://bioconductor.org/packages/release/bioc/html/ggtree.html}{online documentation}.
#'
#' @seealso To generate solutions for the 'Project
#'   Prioritization Protocol' problem, see
#'   \code{\link{ppp_heuristic_phylo_solution}}
#'   \code{\link{ppp_exact_phylo_solution}},
#'   \code{\link{ppp_manual_phylo_solution}},
#'   or \code{\link{ppp_random_phylo_solution}}.
#'
#' @references
#' Faith DP (2008) Threatened species and the potential loss of
#' phylogenetic diversity: conservation scenarios based on estimated extinction
#' probabilities and phylogenetic risk analysis. \emph{Conservation Biology},
#' \strong{22}: 1461--1470.
#'
#' Joseph LN, Maloney RF & Possingham HP (2009) Optimal allocation of
#' resources among threatened species: A project prioritization protocol.
#' \emph{Conservation Biology}, \strong{23}, 328--338.
#'
#' Yu G, Smith DK, Zhu H, Guan Y, & Lam TTY (2017) ggtree: an
#' R package for visualization and annotation of phylogenetic trees with their
#' covariates and other associated data. \emph{Methods in Ecology and
#' Evolution}, \strong{8}: 28--36.
#'
#' @examples
#' # set seed for reproducibility
#' set.seed(500)
#'
#' # load built-in data
#' data(sim_project_data, sim_tree)
#'
#' # load packages to help with plotting
#' library(ggplot2)
#'
#' # print simulated project data set
#' print(sim_project_data)
#'
#' # print simulated phylogenetic tree data set
#' print(sim_tree)
#'
#' # plot the simulated phylogeny
#' plot(sim_tree, main = "simulated phylogeny")
#'
#'
#' solutions <- data.frame(S1_project = c(FALSE, FALSE, TRUE),
#'                         S2_project = c(TRUE, FALSE, TRUE),
#'                         S4_project = c(TRUE, FALSE, TRUE),
#'                         S3_project = c(FALSE, FALSE, TRUE),
#'                         S5_project = c(TRUE, FALSE, TRUE),
#'                         baseline_project = c(TRUE, TRUE, TRUE))
#'
#' print(solutions)
#'
#' # create random some solutions with a budget of 700
#' s1 <- ppp_random_solution(sim_project_data, sim_tree, 700,
#'                           "name", "cost", "success", number_solutions = 10)
#'
#' # print output
#' print(s1)
#'
#' # plot the first solution
#' ppp_plot(sim_project_data, sim_tree, s1, "name", "cost", "success")
#'
#' # plot the second solution
#' ppp_plot(sim_project_data, sim_tree, s1, "name", "cost", "success", n = 2)
#'
#' # since this function returns a ggplot2 plot object, we can customize the
#' # appearance of the plot using standard ggplot2 commands!
#' # for example, we can add a title
#' ppp_plot(sim_project_data, sim_tree, s1, "name", "cost", "success") +
#' ggtitle("solution")
#'
#' # we could also also set the minimum and maximum values in the color ramp to
#' # correspond to those in the data, rather than being capped at 0 and 1
#' ppp_plot(sim_project_data, sim_tree, s1, "name", "cost", "success") +
#' scale_color_gradientn(name = "Probability of\npersistence",
#'                       colors = viridisLite::inferno(150, begin = 0,
#'                                                     end = 0.9,
#'                                                     direction = -1)) +
#' ggtitle("solution")
#'
#' # we could also change the color ramp
#' ppp_plot(sim_project_data, sim_tree, s1, "name", "cost", "success") +
#' scale_color_gradient(name = "Probability of\npersistence",
#'                      low = "red", high = "black") +
#' ggtitle("solution")
#'
#' # we could even hide the legend if desired
#' ppp_plot(sim_project_data, sim_tree, s1, "name", "cost", "success") +
#' scale_color_gradient(name = "Probability of\npersistence",
#'                      low = "red", high = "black") +
#' theme(legend.position = "hide") +
#' ggtitle("solution")
#' @export
ppp_phylo_plot <- function(x, tree, solution, project_column_name,
                           cost_column_name, success_column_name, n = 1L,
                           asterisk_hjust = 0.007) {
  # assertions
  ## assert that ggtree R package is installed
  assertthat::assert_that(requireNamespace("ggtree", quietly = TRUE),
                          msg = "ggtree R package not installed.")
  ## coerce x to tibble if just a regular data.frame
  if (inherits(x, "data.frame") && !inherits(x, "tbl_df"))
    x <- tibble::as_tibble(x)
  ## coerce solution to tibble if just a regular data.frame
  if (inherits(solution, "data.frame") && !inherits(solution, "tbl_df"))
    solution <- tibble::as_tibble(solution)
  ## assert that parameters are valid
  assertthat::assert_that(inherits(x, "tbl_df"),
                          ncol(x) > 0, nrow(x) > 0,
                          inherits(solution, "tbl_df"),
                          ncol(solution) > 0, nrow(solution) > 0,
                          inherits(tree, "phylo"),
                          assertthat::is.string(project_column_name),
                          assertthat::has_name(x, project_column_name),
                          assertthat::noNA(x[[project_column_name]]),
                          inherits(x[[project_column_name]],
                                   c("character", "factor")),
                          assertthat::is.string(cost_column_name),
                          assertthat::has_name(x, cost_column_name),
                          is.numeric(x[[cost_column_name]]),
                          assertthat::noNA(x[[cost_column_name]]),
                          assertthat::is.string(success_column_name),
                          assertthat::has_name(x, success_column_name),
                          is.numeric(x[[success_column_name]]),
                          assertthat::noNA(x[[success_column_name]]),
                          assertthat::is.count(n),
                          is.finite(n),
                          isTRUE(n <= nrow(solution)),
                          assertthat::is.number(asterisk_hjust))
  ## coerce factor species names to character
  if (is.factor(x[[project_column_name]]))
    x[[project_column_name]] <- as.character(x[[project_column_name]])

  ## additional checks
  assertthat::assert_that(all(assertthat::has_name(solution,
                                                   x[[project_column_name]])))
  assertthat::assert_that(all(vapply(solution[, x[[project_column_name]],
                                              drop = FALSE],
                                     class, character(1)) == "logical"))

  # preliminary data processing
  ## check that branches have lengths
  if (is.null(tree$edge.length)) {
    tree$edge.length <- rep(1, nrow(tree$edge))
    warning(paste("tree does not have branch length data,",
                  "all branches are assumed to have equal lengths"))
  }  else {
    assertthat::assert_that(nrow(tree$edge) == length(tree$edge.length))
  }

  ## subset solution and reorder columns
  solution <- solution[n, x[[project_column_name]], drop = FALSE]

  ## determine which species received funding
  funded_spp <- as.matrix(x[, tree$tip.label, drop = FALSE]) > 1e-15
  funded_spp[which(x[[cost_column_name]] < 1e-15), ] <- 0.0
  funded_projects <- as.numeric(as.matrix(solution))
  funded_projects <- matrix(funded_projects, ncol = ncol(funded_spp),
                            nrow = nrow(funded_spp))
  funded_spp <- tree$tip.label[colSums(funded_spp * funded_projects) > 1e-15]

  ## pre-compute conditional probabilities of species persistence
  ## and project success
  spp_probs <- as.matrix(x[, tree$tip.label, drop = FALSE])
  spp_probs <- spp_probs * matrix(x[[success_column_name]],
                                  ncol = ncol(spp_probs),
                                  nrow = nrow(spp_probs))
  spp_probs <- Matrix::drop0(methods::as(round(spp_probs, 5), "dgCMatrix"))

  ## pre-compute probabilities that each branch will persist
  branch_probs <- rcpp_branch_probabilities(spp_probs, branch_matrix(tree),
                                            methods::as(as.matrix(solution),
                                               "dgCMatrix"))

  # Main processing
  ## format tree data for plotting
  tree2 <- suppressMessages(suppressWarnings(tidytree::as_data_frame(tree)))
  tree2$status <- tree2$label %in% funded_spp
  tree2$status <- c("Not Funded", "Funded")[tree2$status + 1]
  tree2$prob <- c(branch_probs)[match(
    paste0(tree2$parent, "_", tree2$node),
    paste0(tree$edge[, 1], "_", tree$edge[, 2]))]
  tree2$label <- paste0("   ", tree2$label)
  tree2 <- suppressMessages(suppressWarnings(tidytree::as.treedata(tree2)))

  ## calculate padding for points
  point_padding <- max(rowSums(as.matrix(branch_matrix(tree)) *
                               matrix(tree$edge.length, ncol = nrow(tree$edge),
                                      nrow = length(tree$tip.label)))) *
                   asterisk_hjust
  ## make plot
  p <- ggtree::ggtree(tree2, ggplot2::aes_string(color = "prob"), size = 1.1) +
       ggtree::geom_tippoint(
         ggplot2::aes_string(subset = "status == \"Funded\"",
                             x = "x + point_padding"),
         color = "black", pch = 8) +
       ggtree::geom_tiplab(color = "black", size = 2.5) +
       ggplot2::scale_color_gradientn(name = "Probability of\npersistence",
                                      colors = viridisLite::inferno(
                                        150, begin = 0, end = 0.9,
                                        direction = -1),
                                      limits = c(0, 1)) +
       ggplot2::theme(legend.position = "right")

  # Exports
  # return plot
  p
}
